File: app/src/main/java/com/xaye/downloader/core/ConnectThread.java
package com.xaye.downloader.core;

import com.xaye.downloader.utilities.Constants;
import com.xaye.downloader.utilities.Trace;

import java.io.IOException;
import java.net.HttpURLConnection;
import java.net.URL;

/**
 * @FileName:com.xaye.downloader.ConnectThread.kt Author xaye
 * @date: 2024-05-04 10:12
 * Created by 11623 on 2024/5/4
 */
public class ConnectThread implements Runnable{

    private final String url;
    private final ConnectListener listener;

    private volatile boolean isRunning;

    public ConnectThread(String url, ConnectListener listener) {
        this.url = url;
        this.listener = listener;
    }

    @Override
    public void run() {
        isRunning = true;
        HttpURLConnection connection = null;
        try{
            connection = (HttpURLConnection) new URL(url).openConnection();
            connection.setRequestMethod("GET");
            //connection.setRequestProperty("Range","bytes=0-"+Integer.MAX_VALUE);
            connection.setConnectTimeout(Constants.CONNECT_TIMEOUT);
            connection.setReadTimeout(Constants.READ_TIMEOUT);
            int responseCode = connection.getResponseCode();
            int contentLength = connection.getContentLength();
            boolean isSupportRange = false;

            if (responseCode == HttpURLConnection.HTTP_OK) {
                String ranges = connection.getHeaderField("Accept-Ranges");
                if (ranges != null && ranges.equals("bytes")) {
                    isSupportRange = true;
                }
                listener.onConnected(isSupportRange,contentLength);

            } else {
                listener.onConnectError("server error responseCode:"+responseCode);
            }

            Trace.d("ConnectThread responseCode:"+responseCode);

            isRunning = false;
        }catch (IOException e) {
            isRunning = false;
            listener.onConnectError(e.getMessage());
            e.printStackTrace();
        } finally {
            if (connection != null) {
                connection.disconnect();
            }
        }
    }

    public boolean isRunning() {
        return isRunning;
    }

    public void cancel() {
        Thread.currentThread().interrupt();
    }

    interface ConnectListener{
        void onConnected(boolean isSupportRange, int totalLength);

        void onConnectError(String message);
    }
}

File: app/src/main/java/com/xaye/downloader/core/ConnectThread.kt
package com.xaye.downloader.core

import com.xaye.downloader.network.DownloadException
import com.xaye.downloader.network.Error
import com.xaye.downloader.network.ExceptionHandle
import com.xaye.downloader.utils.Constants
import com.xaye.downloader.utils.Trace
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL

class ConnectThread(private val url: String, private val listener: ConnectListener) : Runnable {

    @Volatile
    var isRunning = false
        private set

    override fun run() {
        isRunning = true
        var connection: HttpURLConnection? = null
        try {
            connection = URL(url).openConnection() as HttpURLConnection
            connection.requestMethod = "GET"
            connection.connectTimeout = Constants.CONNECT_TIMEOUT
            connection.readTimeout = Constants.READ_TIMEOUT
            val responseCode = connection.responseCode
            val contentLength = connection.contentLength
            var isSupportRange = false

            if (responseCode == HttpURLConnection.HTTP_OK) {
                val ranges = connection.getHeaderField("Accept-Ranges")
                if (ranges != null && ranges == "bytes") {
                    isSupportRange = true
                }
                listener.onConnected(isSupportRange, contentLength)
            } else {
                listener.onConnectError(
                    DownloadException(
                        Error.RESPONSE_CODE_ERROR.getKey(),
                        Error.RESPONSE_CODE_ERROR.getValue(),
                        responseCode.toString(),
                        null
                    )
                )
            }

            Trace.d("ConnectThread responseCode:$responseCode")

            isRunning = false
        } catch (e: IOException) {
            isRunning = false
            Trace.e("ConnectThread error:${e.message}  e: $e")
            listener.onConnectError(ExceptionHandle.handleException(e))
            e.printStackTrace()
        } finally {
            connection?.disconnect()
        }
    }

    fun cancel() {
        Thread.currentThread().interrupt()
    }

    interface ConnectListener {
        fun onConnected(isSupportRange: Boolean, totalLength: Int)

        fun onConnectError(exception: DownloadException)
    }
}
